EventGenerator/src/MuplusMichelGenerator_tool.cc

```diff
@@ -0,0 +1,82 @@
+#include "art/Utilities/ToolMacros.h"
+
+#include "CLHEP/Random/RandPoissonQ.h"
+#include "CLHEP/Random/RandGeneral.h"
+
+#include "Offline/EventGenerator/inc/ParticleGeneratorTool.hh"
+
+#include "Offline/DataProducts/inc/PDGCode.hh"
+#include "Offline/MCDataProducts/inc/GenId.hh"
+#include "Offline/Mu2eUtilities/inc/RandomUnitSphere.hh"
+#include "Offline/Mu2eUtilities/inc/BinnedSpectrum.hh"
+#include "Offline/GlobalConstantsService/inc/GlobalConstantsHandle.hh"
+#include "Offline/GlobalConstantsService/inc/ParticleDataList.hh"
+#include "Offline/GlobalConstantsService/inc/PhysicsParams.hh"
+
+#include "fhiclcpp/types/DelegatedParameter.h"
+
+namespace mu2e {
+  class MuplusMichelGenerator : public ParticleGeneratorTool {
+  public:
+    struct PhysConfig {
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+
+      fhicl::DelegatedParameter spectrum{Name("spectrum"), Comment("Parameters for BinnedSpectrum)")};
+    };
+    typedef art::ToolConfigTable<PhysConfig> Parameters;
+
+    explicit MuplusMichelGenerator(Parameters const& conf) :
+      _pdgId(PDGCode::e_plus),
+      _mass(GlobalConstantsHandle<ParticleDataList>()->particle(_pdgId).mass()),
+      _spectrum(BinnedSpectrum(conf().spectrum.get<fhicl::ParameterSet>()))
+    {}
+
+    std::vector<ParticleGeneratorTool::Kinematic> generate() override;
+    void generate(std::unique_ptr<GenParticleCollection>& out, const IO::StoppedParticleF& stop) override;
+
+    void finishInitialization(art::RandomNumberGenerator::base_engine_t& eng, const std::string&) override {
+      _randomUnitSphere = new RandomUnitSphere(eng);
+      _randSpectrum = new CLHEP::RandGeneral(eng, _spectrum.getPDF(), _spectrum.getNbins());
```

Comment:
This memory leaks.  Please hold these by unique_prt.

Reply:
Could you be more specific because I couldn't understand this comment. Sorry.

Reply:
You can read about memory leaks: https://en.wikipedia.org/wiki/Memory_leak

Make sure that you have this line among your include directives:
#include <memory>

At lines 49 and 50, change the types of the two variables:
std::unique_ptr<RandomUnitSphere> _randomUnitSphere;
std::unique_ptr<CLHEP::RandGeneral> _randSpectrum;

At lines 39 and 40,
_randomUnitSphere = std::make_unique<RandomUnitSphere>(eng);
_randSpectrum = std::make_unique<CLHEP::RandGeneral>( eng, ....);

where you use _randomUnitSphere.  change the . to -> .

This ensures that, when the module destructor is called, the memory used by these two objects will be released.

You can read more at:
https://cplusplus.com/reference/memory/unique_ptr/
https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique

Reply:
Thank you for clarifying this.